### PR TITLE
JCL-309: Specialized HTTP client exceptions

### PR DESCRIPTION
**Pull request by hellikopter**
Changes in PR:
- new specialized exceptions to cover most common HTTP client errors: `400 Bad Request`, `401 Unauthorized`, `403 Forbidden`, `404 Not Found`, `405 Method Not Allowed`, `406 Not Acceptable`, `409 Conflict`, `410 Gone`, `412 Precondition Failed`, `415 Unsupported Media Type`, `429 Too Many Requests`, `500 Internal Server Error`
- `SolidClient` throws specialized exceptions whenever applicable

### CODE REVIEW

File: solid/src/main/java/com/inrupt/client/solid/BadRequestException.java [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+ * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+package com.inrupt.client.solid;
+
+import com.inrupt.client.Headers;
+
+import java.net.URI;
+
+/**
+ * A runtime exception for use with SolidClient HTTP operations.
```
Comment (langsamu) [inline solid/src/main/java/com/inrupt/client/solid/BadRequestException.java:28]:
```suggestion
 * A runtime exception that represents an HTTP bad request (400) response.
 *
 * @see <a href="https://www.rfc-editor.org/rfc/rfc9110#status.400">RFC 9110 (15.5.1.) 400 Bad Request</a>
```

File: solid/src/main/java/com/inrupt/client/solid/ConflictException.java [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+ * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+package com.inrupt.client.solid;
+
+import com.inrupt.client.Headers;
+
+import java.net.URI;
+
+/**
+ * A runtime exception for use with SolidClient HTTP operations.
```
Comment (langsamu) [inline solid/src/main/java/com/inrupt/client/solid/ConflictException.java:28]:
```suggestion
 * A runtime exception that represents an HTTP conflict (409) response..
 *
 * @see <a href="https://www.rfc-editor.org/rfc/rfc9110#status.409">RFC 9110 (15.5.10.) 409 Conflict</a>
```

File: solid/src/main/java/com/inrupt/client/solid/ForbiddenException.java [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+ * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+package com.inrupt.client.solid;
+
+import com.inrupt.client.Headers;
+
+import java.net.URI;
+
+/**
+ * A runtime exception for use with SolidClient HTTP operations.
```
Comment (langsamu) [inline solid/src/main/java/com/inrupt/client/solid/ForbiddenException.java:28]:
```suggestion
 * A runtime exception that represents an HTTP forbidden (403) response.
 *
 * @see <a href="https://www.rfc-editor.org/rfc/rfc9110#status.403">RFC 9110 (15.5.4.) 403 Forbidden</a>
```

File: solid/src/main/java/com/inrupt/client/solid/GoneException.java [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+ * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+package com.inrupt.client.solid;
+
+import com.inrupt.client.Headers;
+
+import java.net.URI;
+
+/**
+ * A runtime exception for use with SolidClient HTTP operations.
```
Comment (langsamu) [inline solid/src/main/java/com/inrupt/client/solid/GoneException.java:28]:
```suggestion
 * A runtime exception that represents an HTTP gone (410) response.
 *
 * @see <a href="https://www.rfc-editor.org/rfc/rfc9110#status.410">RFC 9110 (15.5.11.) 410 Gone</a>
```

File: solid/src/main/java/com/inrupt/client/solid/InternalServerErrorException.java [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+ * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+package com.inrupt.client.solid;
+
+import com.inrupt.client.Headers;
+
+import java.net.URI;
+
+/**
+ * A runtime exception for use with SolidClient HTTP operations.
```
Comment (langsamu) [inline solid/src/main/java/com/inrupt/client/solid/InternalServerErrorException.java:28]:
```suggestion
 * A runtime exception that represents an HTTP internal server error (500) response.
 *
 * @see <a href="https://www.rfc-editor.org/rfc/rfc9110#status.500">RFC 9110 (15.6.1.) 500 Internal Server Error</a>
```

File: solid/src/main/java/com/inrupt/client/solid/MethodNotAllowedException.java [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+ * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+package com.inrupt.client.solid;
+
+import com.inrupt.client.Headers;
+
+import java.net.URI;
+
+/**
+ * A runtime exception for use with SolidClient HTTP operations.
```
Comment (langsamu) [inline solid/src/main/java/com/inrupt/client/solid/MethodNotAllowedException.java:28]:
```suggestion
 * A runtime exception that represents an HTTP method not allowed (405) response.
 *
 * @see <a href="https://www.rfc-editor.org/rfc/rfc9110#status.405">RFC 9110 (15.5.6.) 405 Method Not Allowed</a>
```

File: solid/src/main/java/com/inrupt/client/solid/NotAcceptableException.java [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+ * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+package com.inrupt.client.solid;
+
+import com.inrupt.client.Headers;
+
+import java.net.URI;
+
+/**
+ * A runtime exception for use with SolidClient HTTP operations.
```
Comment (langsamu) [inline solid/src/main/java/com/inrupt/client/solid/NotAcceptableException.java:28]:
```suggestion
 * A runtime exception that represents an HTTP not acceptable (406) response.
 *
 * @see <a href="https://www.rfc-editor.org/rfc/rfc9110#status.406">RFC 9110 (15.5.7.) 406 Not Acceptable</a>
```

File: solid/src/main/java/com/inrupt/client/solid/NotFoundException.java [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+ * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+package com.inrupt.client.solid;
+
+import com.inrupt.client.Headers;
+
+import java.net.URI;
+
+/**
+ * A runtime exception for use with SolidClient HTTP operations.
```
Comment (langsamu) [inline solid/src/main/java/com/inrupt/client/solid/NotFoundException.java:28]:
```suggestion
 * A runtime exception that represents an HTTP not found (404) response.
 *
 * @see <a href="https://www.rfc-editor.org/rfc/rfc9110#status.404">RFC 9110 (15.5.5.) 404 Not Found</a>
```

File: solid/src/main/java/com/inrupt/client/solid/PreconditionFailedException.java [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+ * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+package com.inrupt.client.solid;
+
+import com.inrupt.client.Headers;
+
+import java.net.URI;
+
+/**
+ * A runtime exception for use with SolidClient HTTP operations.
```
Comment (langsamu) [inline solid/src/main/java/com/inrupt/client/solid/PreconditionFailedException.java:28]:
```suggestion
 * A runtime exception that represents an HTTP precondition failed (412) response.
 *
 * @see <a href="https://www.rfc-editor.org/rfc/rfc9110#status.412">RFC 9110 (15.5.13.) 412 Precondition Failed</a>
```

File: solid/src/main/java/com/inrupt/client/solid/TooManyRequestsException.java [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+ * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+package com.inrupt.client.solid;
+
+import com.inrupt.client.Headers;
+
+import java.net.URI;
+
+/**
+ * A runtime exception for use with SolidClient HTTP operations.
```
Comment (langsamu) [inline solid/src/main/java/com/inrupt/client/solid/TooManyRequestsException.java:28]:
```suggestion
 * A runtime exception that represents an HTTP too many requests (429) response.
 *
 * @see <a href="https://www.rfc-editor.org/rfc/rfc6585#section-4">RFC 6585 (4.) 429 Too Many Requests</a>
```

File: solid/src/main/java/com/inrupt/client/solid/UnauthorizedException.java [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+ * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+package com.inrupt.client.solid;
+
+import com.inrupt.client.Headers;
+
+import java.net.URI;
+
+/**
+ * A runtime exception for use with SolidClient HTTP operations.
```
Comment (langsamu) [inline solid/src/main/java/com/inrupt/client/solid/UnauthorizedException.java:28]:
```suggestion
 * A runtime exception that represents an HTTP unauthorized (401) response.
 *
 * @see <a href="https://www.rfc-editor.org/rfc/rfc9110#status.401">RFC 9110 (15.5.2.) 401 Unauthorized</a>
```

File: solid/src/main/java/com/inrupt/client/solid/UnsupportedMediaTypeException.java [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+ * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+package com.inrupt.client.solid;
+
+import com.inrupt.client.Headers;
+
+import java.net.URI;
+
+/**
+ * A runtime exception for use with SolidClient HTTP operations.
```
Comment (langsamu) [inline solid/src/main/java/com/inrupt/client/solid/UnsupportedMediaTypeException.java:28]:
```suggestion
 * A runtime exception that represents an HTTP unsupported media type (415) response.
 *
 * @see <a href="https://www.rfc-editor.org/rfc/rfc9110#status.415">RFC 9110 (15.5.16.) 415 Unsupported Media Type</a>
```

**Review by langsamu — CHANGES_REQUESTED** [review]

**Review by langsamu — APPROVED** [review]

File: solid/src/test/java/com/inrupt/client/solid/SolidClientTest.java [hunk 13]
```diff
@@ -243,14 +241,18 @@ void testGetRecipeType() {
...
+            final int expectedStatusCode,
+            final Class<T> clazz) {
 
         final CompletableFuture<Recipe> future =
             client.read(uri, Recipe.class)
             .toCompletableFuture();
         final CompletionException err = assertThrows(CompletionException.class, () ->
             future.join());
         assertTrue(err.getCause() instanceof SolidClientException);
+        assertInstanceOf(clazz, err.getCause());
```
Comment (acoburn) [inline solid/src/test/java/com/inrupt/client/solid/SolidClientTest.java:255]: This is nice -- I'll need to make more use of this method

**Review by acoburn — APPROVED** [review]
LGTM

**Pull request merged**